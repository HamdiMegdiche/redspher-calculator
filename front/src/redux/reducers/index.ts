import { combineReducers } from "redux";
import calculator from "./calculator";

/**
 * Combined reducers of the redux store
 */
const reducers = combineReducers({
  calculator
});

export default reducers;
