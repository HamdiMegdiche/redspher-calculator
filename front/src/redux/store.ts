import { configureStore } from "@reduxjs/toolkit";
import rootReducers from "./reducers";
// Store configuration
export const store = configureStore({
  reducer: rootReducers
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
