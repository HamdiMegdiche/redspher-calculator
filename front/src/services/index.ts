import axios from "axios";

export async function computeString(text: string): Promise<[string | null, any]> {
  try {
    const url = process.env.REACT_APP_API_URL + "/compute";
    const response = await axios.post(url, { text });
    return [response.data, null];
  } catch (e) {
    console.error(e);
    return [null, e];
  }
}
