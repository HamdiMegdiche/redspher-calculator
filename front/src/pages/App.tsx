import { useEffect, useState } from "react";
import styled from "styled-components";
import { useAppDispatch, useAppSelector } from "redux/hooks";
import { hasOperations } from "helpers";
import { Button } from "components";
import { actions } from "redux/reducers/calculator";
import { computeString } from "services";

export default function App() {
  const dispatch = useAppDispatch();
  const [operation, setOperation] = useState("");
  const [previousOperationText, setPreviousOperationText] = useState<string | null>(null);
  // Operations list selector from redux store
  const operations = useAppSelector(state => state.calculator.operations);
  const buttons = ["/", "1", "2", "3", "*", "4", "5", "6", "+", "7", "8", "9", "-", ".", "0"];

  useEffect(() => {
    window.addEventListener("keydown", handleEventKeyDownCallback);
    // CLEANUP
    return () => {
      window.removeEventListener("keydown", handleEventKeyDownCallback);
    };
  }, [operation]);

  function clear() {
    setOperation("");
    setPreviousOperationText(null);
    dispatch(actions.clearOperations());
  }

  async function handleEventKeyDownCallback(e: KeyboardEvent) {
    e.stopImmediatePropagation();
    const keys = [...buttons, "Delete", "Backspace", "=", "Enter"];
    if (!keys.includes(e.key)) return;

    if (["Delete", "Backspace"].includes(e.key)) handleDelete();
    else if (["=", "Enter"].includes(e.key)) await compute();
    else setOperation(v => (v += e.key));
  }

  function handleDelete() {
    if (!operation.length) return;
    const newValue = operation.substring(0, operation.length - 1);
    setOperation(newValue);
  }

  async function compute() {
    if (!operation.length || !hasOperations(operation)) return;

    const [response, error] = await computeString(operation);
    if (error) {
      console.error(error);
      return;
    }
    dispatch(actions.addOperation({ value: operation, computed: response! }));
    setPreviousOperationText(`${operation}=${response}`);
    setOperation("");
  }

  return (
    <Wrapper>
      <Container>
        <Output>
          <Previous data-testid="previous-operation">{previousOperationText ?? "ANS"}</Previous>
          <Current data-testid="current-operation">{operation.length ? operation : 0}</Current>
        </Output>
        <ButtonStyled span="2" label="AC" backgroundColor="#dadce0" onClick={clear} />
        <ButtonStyled label="DEL" backgroundColor="#dadce0" onClick={handleDelete} />

        {buttons.map((b, index) => (
          <ButtonStyled
            key={`${b}-${index}`}
            label={b}
            backgroundColor={hasOperations(b) ? "#dadce0" : "#f1f3f4"}
            onClick={() => setOperation(v => (v += b))}
          />
        ))}

        <ButtonStyled
          span="2"
          label="="
          backgroundColor="#4285f4"
          color="white"
          onClick={compute}
        />
      </Container>
    </Wrapper>
  );
}

const Wrapper = styled.div`
  height: 100vh;
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: center;
  background-color: white;
`;

const Container = styled.div`
  display: grid;
  grid-template-columns: repeat(4, 85px);
  grid-template-rows: repeat(7, 40px);
  gap: 2px;
`;

const Output = styled.div`
  border: 1px grey solid;
  border-radius: 10px;
  margin: 4px;
  grid-column: span 4;
  grid-row: span 2;
  font-family: monospace;
  grid-column: 1 / -1;
  display: flex;
  align-items: flex-end;
  justify-content: space-around;
  flex-direction: column;
  padding: 8px;
  word-wrap: break-word;
  word-break: break-all;
  overflow: auto;
`;

const Previous = styled.span`
  font-size: 16px;
  color: #70757a;
`;
const Current = styled.span`
  font-size: 30px;
`;
const ButtonStyled = styled(Button)`
  user-select: none;
  outline: none;
  border-radius: 4px;
  margin: 4px;
`;
