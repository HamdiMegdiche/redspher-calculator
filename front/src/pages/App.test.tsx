import { fireEvent, render, RenderResult, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "redux/store";
import App from "./App";

import * as services from "services";
import { act } from "react-dom/test-utils";
const mockCompute = jest.spyOn(services, "computeString");

const renderApp = (): RenderResult =>
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );

it("Renders properly", async () => {
  const { asFragment } = renderApp();
  expect(asFragment()).toMatchSnapshot();
  const prevOperation = await screen.findByText("ANS");
  expect(prevOperation).toBeInTheDocument();
  // There is 2 zeros : a button and current operation
  const zeroElements = await screen.findAllByText("0");
  expect(zeroElements).toHaveLength(2);
  expect(store.getState().calculator.operations).toEqual([]);
});

it("Current operation correctly changes after click", async () => {
  const { asFragment } = renderApp();
  expect(asFragment()).toMatchSnapshot();
  const button1 = await screen.getByText("1");
  const buttonPlus = await screen.getByText("+");
  const button2 = await screen.getByText("2");
  fireEvent.click(button1);
  fireEvent.click(buttonPlus);
  fireEvent.click(button2);
  const currentOperation = await screen.getByTestId("current-operation");
  expect(currentOperation.textContent).toEqual("1+2");
});

it("Current operation correctly changes after keydown", async () => {
  const { asFragment } = renderApp();
  expect(asFragment()).toMatchSnapshot();
  fireEvent.keyDown(document, { key: "1" });
  fireEvent.keyDown(document, { key: "+" });
  fireEvent.keyDown(document, { key: "2" });
  fireEvent.keyDown(document, { key: "+" });
  fireEvent.keyDown(document, { key: "Backspace" });
  const currentOperation = screen.getByTestId("current-operation");
  expect(currentOperation.textContent).toEqual("1+2");
});

it("Current operation correctly changes after Clear", async () => {
  const { asFragment } = renderApp();
  expect(asFragment()).toMatchSnapshot();
  const buttonAC = await screen.getByText("AC");
  fireEvent.keyDown(document, { key: "1" });
  fireEvent.keyDown(document, { key: "+" });
  fireEvent.keyDown(document, { key: "2" });
  fireEvent.keyDown(document, { key: "2" });
  fireEvent.click(buttonAC);
  const currentOperation = screen.getByTestId("current-operation");
  expect(currentOperation.textContent).toEqual("0");
  expect(store.getState().calculator.operations).toEqual([]);
});

it("Checks Previous & Current operations correctly changes after compute API Call", async () => {
  const { asFragment } = renderApp();
  await mockCompute.mockResolvedValue(["3", null]);
  expect(asFragment()).toMatchSnapshot();
  fireEvent.keyDown(document, { key: "1" });
  fireEvent.keyDown(document, { key: "+" });
  fireEvent.keyDown(document, { key: "2" });
  const currentOperation = await screen.getByTestId("current-operation");
  expect(currentOperation.textContent).toEqual("1+2");
  act(async () => {
    fireEvent.keyDown(document, { key: "Enter" });
    const previousOperation = await screen.getByTestId("previous-operation");
    expect(previousOperation.textContent).toEqual("1+2=3");
    expect(currentOperation.textContent).toEqual("0");
    const operations = store.getState().calculator.operations;
    expect(operations).toEqual([{ value: "1+2", computed: "3" }]);
  });
});
