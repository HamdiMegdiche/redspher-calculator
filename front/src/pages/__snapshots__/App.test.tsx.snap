// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checks Previous & Current operations correctly changes after compute API Call 1`] = `
<DocumentFragment>
  <div
    class="sc-gtsrHT dfpLLh"
  >
    <div
      class="sc-dlnjwi iChkeo"
    >
      <div
        class="sc-hKFxyN fBuZAB"
      >
        <span
          class="sc-eCApnc bHRJIF"
          data-testid="previous-operation"
        >
          ANS
        </span>
        <span
          class="sc-jSFjdj bVZAPG"
          data-testid="current-operation"
        >
          0
        </span>
      </div>
      <button
        class="sc-bdnxRM rpSRV sc-gKAaRy kbkzzC"
        color="black"
        label="AC"
        span="2"
      >
        AC
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="DEL"
      >
        DEL
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="/"
      >
        /
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="1"
      >
        1
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="2"
      >
        2
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="3"
      >
        3
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="*"
      >
        *
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="4"
      >
        4
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="5"
      >
        5
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="6"
      >
        6
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="+"
      >
        +
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="7"
      >
        7
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="8"
      >
        8
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="9"
      >
        9
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="-"
      >
        -
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="."
      >
        .
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="0"
      >
        0
      </button>
      <button
        class="sc-bdnxRM eJuZHN sc-gKAaRy kbkzzC"
        color="white"
        label="="
        span="2"
      >
        =
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Current operation correctly changes after Clear 1`] = `
<DocumentFragment>
  <div
    class="sc-gtsrHT dfpLLh"
  >
    <div
      class="sc-dlnjwi iChkeo"
    >
      <div
        class="sc-hKFxyN fBuZAB"
      >
        <span
          class="sc-eCApnc bHRJIF"
          data-testid="previous-operation"
        >
          ANS
        </span>
        <span
          class="sc-jSFjdj bVZAPG"
          data-testid="current-operation"
        >
          0
        </span>
      </div>
      <button
        class="sc-bdnxRM rpSRV sc-gKAaRy kbkzzC"
        color="black"
        label="AC"
        span="2"
      >
        AC
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="DEL"
      >
        DEL
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="/"
      >
        /
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="1"
      >
        1
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="2"
      >
        2
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="3"
      >
        3
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="*"
      >
        *
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="4"
      >
        4
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="5"
      >
        5
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="6"
      >
        6
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="+"
      >
        +
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="7"
      >
        7
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="8"
      >
        8
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="9"
      >
        9
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="-"
      >
        -
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="."
      >
        .
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="0"
      >
        0
      </button>
      <button
        class="sc-bdnxRM eJuZHN sc-gKAaRy kbkzzC"
        color="white"
        label="="
        span="2"
      >
        =
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Current operation correctly changes after click 1`] = `
<DocumentFragment>
  <div
    class="sc-gtsrHT dfpLLh"
  >
    <div
      class="sc-dlnjwi iChkeo"
    >
      <div
        class="sc-hKFxyN fBuZAB"
      >
        <span
          class="sc-eCApnc bHRJIF"
          data-testid="previous-operation"
        >
          ANS
        </span>
        <span
          class="sc-jSFjdj bVZAPG"
          data-testid="current-operation"
        >
          0
        </span>
      </div>
      <button
        class="sc-bdnxRM rpSRV sc-gKAaRy kbkzzC"
        color="black"
        label="AC"
        span="2"
      >
        AC
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="DEL"
      >
        DEL
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="/"
      >
        /
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="1"
      >
        1
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="2"
      >
        2
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="3"
      >
        3
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="*"
      >
        *
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="4"
      >
        4
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="5"
      >
        5
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="6"
      >
        6
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="+"
      >
        +
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="7"
      >
        7
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="8"
      >
        8
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="9"
      >
        9
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="-"
      >
        -
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="."
      >
        .
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="0"
      >
        0
      </button>
      <button
        class="sc-bdnxRM eJuZHN sc-gKAaRy kbkzzC"
        color="white"
        label="="
        span="2"
      >
        =
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Current operation correctly changes after keydown 1`] = `
<DocumentFragment>
  <div
    class="sc-gtsrHT dfpLLh"
  >
    <div
      class="sc-dlnjwi iChkeo"
    >
      <div
        class="sc-hKFxyN fBuZAB"
      >
        <span
          class="sc-eCApnc bHRJIF"
          data-testid="previous-operation"
        >
          ANS
        </span>
        <span
          class="sc-jSFjdj bVZAPG"
          data-testid="current-operation"
        >
          0
        </span>
      </div>
      <button
        class="sc-bdnxRM rpSRV sc-gKAaRy kbkzzC"
        color="black"
        label="AC"
        span="2"
      >
        AC
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="DEL"
      >
        DEL
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="/"
      >
        /
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="1"
      >
        1
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="2"
      >
        2
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="3"
      >
        3
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="*"
      >
        *
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="4"
      >
        4
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="5"
      >
        5
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="6"
      >
        6
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="+"
      >
        +
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="7"
      >
        7
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="8"
      >
        8
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="9"
      >
        9
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="-"
      >
        -
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="."
      >
        .
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="0"
      >
        0
      </button>
      <button
        class="sc-bdnxRM eJuZHN sc-gKAaRy kbkzzC"
        color="white"
        label="="
        span="2"
      >
        =
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Renders properly 1`] = `
<DocumentFragment>
  <div
    class="sc-gtsrHT dfpLLh"
  >
    <div
      class="sc-dlnjwi iChkeo"
    >
      <div
        class="sc-hKFxyN fBuZAB"
      >
        <span
          class="sc-eCApnc bHRJIF"
          data-testid="previous-operation"
        >
          ANS
        </span>
        <span
          class="sc-jSFjdj bVZAPG"
          data-testid="current-operation"
        >
          0
        </span>
      </div>
      <button
        class="sc-bdnxRM rpSRV sc-gKAaRy kbkzzC"
        color="black"
        label="AC"
        span="2"
      >
        AC
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="DEL"
      >
        DEL
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="/"
      >
        /
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="1"
      >
        1
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="2"
      >
        2
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="3"
      >
        3
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="*"
      >
        *
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="4"
      >
        4
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="5"
      >
        5
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="6"
      >
        6
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="+"
      >
        +
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="7"
      >
        7
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="8"
      >
        8
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="9"
      >
        9
      </button>
      <button
        class="sc-bdnxRM cOanBT sc-gKAaRy kbkzzC"
        color="black"
        label="-"
      >
        -
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="."
      >
        .
      </button>
      <button
        class="sc-bdnxRM joDfND sc-gKAaRy kbkzzC"
        color="black"
        label="0"
      >
        0
      </button>
      <button
        class="sc-bdnxRM eJuZHN sc-gKAaRy kbkzzC"
        color="white"
        label="="
        span="2"
      >
        =
      </button>
    </div>
  </div>
</DocumentFragment>
`;
